lib/common/patchset.c

```diff
@@ -41,6 +41,14 @@ add_xml_changes_to_patchset(xmlNode *xml, xmlNode *patchset)
     xml_node_private_t *nodepriv = xml->_private;
     const char *value = NULL;
 
+    if (nodepriv == NULL) {
+        /* Elements that shouldn't occur in a CIB don't have _private set. They
+         * should be stripped out, ignored, or have an error thrown by any code
+         * that processes their parent, so we ignore any changes to them.
+         */
+        return;
+    }
```

Comment:
Users can disable schema validation, so we do need to make sure behavior is reasonable for anything that might show up. Pacemaker ignores text nodes but I don't believe there's any prohibition against them even in the schema; what happens if someone adds superfluous text to an element, then later edits or removes it? Do we ignore it from the beginning, or what?

Reply:
Ignore it from the beginning, basically. `filename2xml()` and `cib_perform_op()` (called during any replacement or modification) both strip text nodes from the in-memory CIB... though in the case of `cib_perform_op()`, this happens after `xml_calculate_changes()`, interestingly. The good news is it happens *before* creating the patchset.

So I'm not sure that scenario is actually a problem. Let me know what you think.

---

However, I do need to make sure we handle v1 patchsets properly.

Reply:
That sounds fine

Reply:
BTW, for nodes *other* than text nodes that don't have private data (e.g., `XML_DTD_NODE`, `XML_CDATA_SECTION_NODE`), Pacemaker doesn't strip them, and we would still be ignoring them in the patchset. The task was to handle text nodes gracefully in patchsets, so I didn't consider other types. I don't want to worry about it (it would be more complicated) but maybe we should.

Before this commit it looks like we'd segfault on CDATA (dereferencing a null pointer in `add_xml_changes_to_patchset()`). It looks like we never use CDATA in the CIB, so that's why it hasn't happened. After this commit, we'd simply ignore CDATA when creating patchsets.

Reply:
> However, I do need to make sure we handle v1 patchsets properly.

`subtract_xml_object()` is the only v1 function that uses XML private data, and it null-checks before dereferencing.